pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+# Code intelligence
+
+We'd like to replace elixir_sense with our own implmentation for the following bits
+of code intelligence.
+
+ * Go to definition
+ * Find references
+ * Refactoring
+
+In order to have this level of intelligence of elixir code, we need to
+search both beam files and source code to discover relevant parts and
+locations of elixir and erlang code, index them an make them
+searchable. This is a large task that will not be accomplished in one
+commit, and this document will help to lay out the scope of the work,
+as well as collect thoughts and drive the implementation.
+
+Code intelligence will be oragnized into the following discrete areas:
+
+* **Discovery** - Finding points of interest in the project's codebase, dependencies and elixir core modules
+* **Caching** - Writing the discovered items to disk and ensuring the index is up to date with the state of the project at *all times*. The goal here is to never have to delete the `.lexical` directory.
```

Comment:
I wonder if we are going to need git hooks for situations where `git checkout <branch>`.
Probably better than "watching" _all_ the files for changes.
Or perhaps we "watch" only the `.git` directory for changes?

Reply:
Hm another situation is that editing files in the editor is something that can be easily communicated to lexical, but file edits could come from other places like `rm <file>` or an editor without lsp configuration.

Reply:
> I wonder if we are going to need git hooks

We won't need "git hooks", but we will need to pay attention to file changes (we already have support for them right now) and re-index files that change due to checkouts.

> file edits could come from other places like
Yes, that's one of the reasons why having reindexing jobs that pay attention to filesystem changes is so important.